rust/src/lib.rs

```diff
@@ -184,24 +186,85 @@ pub fn add_to_linker<T: 'static>(
+                    let name_reader =
+                        wasmparser::NameSectionReader::new(custom.data(), custom.data_offset());
+                    for x in name_reader.into_iter() {
+                        if let wasmparser::Name::Function(f) = x? {
+                            for k in f.into_iter() {
+                                let k = k?;
+                                function_names.insert(k.index, k.name.to_string());
+                            }
                         }
                     }
+                    continue;
                 }
-                continue;
             }
+            wasmparser::Payload::GlobalSection(globalsec) => {
+                for (i, aglob) in globalsec.into_iter().enumerate() {
+                    let glob = aglob.unwrap();
+                    if glob.ty.content_type != wasmparser::ValType::I32 {
+                        continue;
+                    }
+                    let mut reader = glob.init_expr.get_binary_reader();
+                    let opcode = match reader.read_u8() {
+                        Ok(opcode) => opcode,
+                        Err(_) => continue,
+                    };
+                    // i32.const
+                    if opcode != 0x41 {
+                        continue;
+                    }
+                    // due to binaryen limitations u32 version values are encoded as a signed LEB128
+                    // integers so they must be casted back to unsigned
+                    let iv = reader.read_var_i32().unwrap();
+                    let uv = iv as u32;
+                    globals.insert(i as u32, uv);
+                }
+            }
+            wasmparser::Payload::ExportSection(exportsec) => {
+                for export in exportsec.into_iter() {
+                    let export = export?;
+                    if export.kind != wasmparser::ExternalKind::Global {
+                        continue;
+                    }
+                    match export.name {
+                        "wasm_instr_version_major" => {
+                            maj_index = Some(export.index);
+                        }
+                        "wasm_instr_version_minor" => {
+                            min_index = Some(export.index);
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            _ => (),
+        }
+    }
+    // For now tolerate version info not being present
+    let version_info_present = maj_index.is_some() || min_index.is_some();
+    if version_info_present {
+        let maj_index = maj_index.unwrap();
+        let min_index = min_index.unwrap();
+        let maj_num: u32 = *globals.get(&maj_index).unwrap();
+        let min_num: u32 = *globals.get(&min_index).unwrap();
+        if maj_num != WASM_INSTR_VERSION_MAJOR {
+            panic!("wasm wasm-instr major version {maj_num} is not equal to {WASM_INSTR_VERSION_MAJOR}!")
```

Comment:
We probably don't want to panic here, i think you could use `anyhow::bail!` instead